Patch driver.py to allow more flexible sampling

kilojoules opened the pull request:
This patch is useful for sampling OpenMDAO models without treating certain stochastic variables as "actual" design variables. This was developed alongside NREL's DAKOTA OpenMDAO driver https://github.com/WISDEM/dakota_driver. This allows sampling of arrays by index (for array foo I can now sample foo[i]).

Kenneth-T-Moore commented on openmdao/core/driver.py:
```diff
@@ -474,7 +475,14 @@ def set_desvar(self, name, value):
+            vn =re.findall("(.*)\[(.*)\]", name)[0][0] 
+            nv = int(re.findall("(.*)\[(.*)\]", name)[0][1])
+            self.root.unknowns[vn][nv] = value
+            return
```
Independent of the other comment, this `return` is too early, as it misses the code that applies `scaler` and `adder` to the design variable. The `scaler` and `adder` are arguments to `add_desvar` that allow the driver to see a transformed version of this variable for better numerical stability.

kilojoules replied:
Good point. I'll take another look at this.

Kenneth-T-Moore commented:
We also will need a test for this; the tests are important so that we don't ever break elements of the API. Someone on our team can help you with this if you need.

Kenneth-T-Moore commented:
Ok, I think I understand your explanation now, and I think the best way to do this is to add a couple of new methods to the DAKOTA_Driver API:

`add_sampling_variable`
`set_sampling_variable`

that do what you need. I don't think this should be in `Driver` because it violates the premise that the driver sees and controls the model through what the user declares as des_vars, objectives, and constraints.

kilojoules commented:
@Kenneth-T-Moore I moved away from regular expressions and added a index option to set_desvar in b947956b64b64cab2186112e4994f9c94ba0dbee. Please let me know what you think.

Kenneth-T-Moore commented:
This looks good now. It still needs a test, though now that I understand the purpose, I could just pull it down and add a test.

kilojoules commented:
Great! Thank you.

Best,
Julian

On Fri, Oct 28, 2016 at 7:32 AM, Kenneth Moore <notifications@github.com>
wrote:

> This looks good now. It still needs a test, though now that I understand
> the purpose, I could just pull it down and add a test.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMDAO/OpenMDAO/pull/650#issuecomment-256920808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHqP7kOyZAy5PmMFYf2eVobzxlH5q3vtks5q4flVgaJpZM4Kbe02>
> .
>
